### PR TITLE
To fix #2583

### PR DESCRIPTION
**Pull request by plantain-00**
https://github.com/Microsoft/TypeScript/issues/2583

It seems the constructor of interface URL is not generated, is this a bug?

### CODE REVIEW

**Comment by mhegazy** [comment]
that is not the correct change. this adds a new type called `URLConstructor`. what you want is to enable `URL` to be called with `new`. 

the desired output is:
```ts
interface URL{
    hash: string;
    search: string;
    pathname: string;
    port: string;
    hostname: string;
    host: string;
    password: string;
    username: string;
    protocol: string;
    origin: string;
    href: string;
}
declare var URL: {
    revokeObjectURL(url: string): void;
    createObjectURL(object: any, options?: ObjectURLOptions): string;
    new(url: string, base?: string): URL;
};
```

**Comment by plantain-00** [comment]
Did I miss something? Locally I added these into `overridingTypes.json` or `addedTypes.json`, but nothings changes:
```
    {
        "kind": "constructor",
        "interface": "URL",
        "signatures": [
            "new (url: string, base?: string): URL"
        ]
    },
    {
        "kind": "property",
        "interface": "URL",
        "name": "hash",
        "type": "number"
    },
    {
        "kind": "property",
        "interface": "URL",
        "name": "search",
        "type": "number"
    },
    {
        "kind": "property",
        "interface": "URL",
        "name": "pathname",
        "type": "number"
    },
    {
        "kind": "property",
        "interface": "URL",
        "name": "port",
        "type": "number"
    },
    {
        "kind": "property",
        "interface": "URL",
        "name": "hostname",
        "type": "number"
    },
    {
        "kind": "property",
        "interface": "URL",
        "name": "host",
        "type": "number"
    },
    {
        "kind": "property",
        "interface": "URL",
        "name": "password",
        "type": "number"
    },
    {
        "kind": "property",
        "interface": "URL",
        "name": "username",
        "type": "number"
    },
    {
        "kind": "property",
        "interface": "URL",
        "name": "protocol",
        "type": "number"
    },
    {
        "kind": "property",
        "interface": "URL",
        "name": "origin",
        "type": "number"
    },
    {
        "kind": "property",
        "interface": "URL",
        "name": "href",
        "type": "number"
    }
```
I also put these into `overridingTypes.json`, nothing changes too:
```
    {
        "kind": "interface",
        "name": "URL",
        "properties": [
            {
                "name": "hash",
                "type": "string"
            },
            {
                "name": "search",
                "type": "string"
            },
            {
                "name": "pathname",
                "type": "string"
            },
            {
                "name": "port",
                "type": "string"
            },
            {
                "name": "hostname",
                "type": "string"
            },
            {
                "name": "host",
                "type": "string"
            },
            {
                "name": "password",
                "type": "string"
            },
            {
                "name": "username",
                "type": "string"
            },
            {
                "name": "protocol",
                "type": "string"
            },
            {
                "name": "origin",
                "type": "string"
            },
            {
                "name": "href",
                "type": "string"
            }
        ]
    }
```

**Comment by mhegazy** [comment]
@zhengbli ideas?

**Comment by zhengbli** [comment]
This is because an issue in the script. Because the `URL` interface is marked as a static interface in the XML file, the script doesn't pick up added constructors. I'll fix soon.

**Comment by zhengbli** [comment]
@plantain-00 Now adding the content to `addedTypes.json` should do the right thing. Can you give it a try? Thanks

**Comment by plantain-00** [comment]
The tests fails, but `build.cmd` works fine locally.

Also I have no idea how to change the type of variable `URL` to 
```typescript
{
    revokeObjectURL(url: string): void;
    createObjectURL(object: any, options?: ObjectURLOptions): string;
    new(url: string, base?: string): URL;
}
```
I tried to add the following into `addedTypes.json` or `overridingTypes.json`, the result is it added it into all interfaces, or changed the `URL` property in interface `Document`：
```json
{
        "kind": "property",
        "name": "URL",
        "type": "any"
    }
```
So need help.

File: inputfiles/removedTypes.json [hunk 1]
```diff
@@ -21,6 +21,16 @@
     },
     {
         "kind": "method",
+        "interface": "URL",
+        "name": "createObjectURL"
```
Comment (mhegazy) [inline inputfiles/removedTypes.json]: these should be added on a new interface `URLConstructor`, and `var URL: URLConstructor`

**Comment by plantain-00** [comment]
Error appears when run `npm install` in Typescript. See the last build: https://ci.appveyor.com/project/zhengbli/tsjs-lib-generator/build/1.0.78-oncrkjps .
The message is:
```
npm ERR! peerinvalid Peer tslint@3.3.0-dev.1 wants typescript@>=1.7.3 || >=1.7.0-dev.20151003 || >=1.8.0-dev
```

**Comment by mhegazy** [comment]
thanks. @vladima is fixing this.

**Comment by plantain-00** [comment]
Can not rerun the failed build, so restart a new on at #58

**Pull request closed without merging**